lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/Gen.hs

```diff
@@ -7,28 +11,258 @@
 --
 module Internal.Cardano.Write.Tx.Gen
     ( genDatumHash
+    , mockPParams
     )
     where
 
 import Prelude
 
+import Cardano.Api.Ledger
+    ( Coin (..)
+    , CostModels
+    , PParams
+    , PParamsHKD
+    , UpgradeConwayPParams (..)
+    )
 import Data.Maybe
     ( fromMaybe
     )
 import Internal.Cardano.Write.Tx
-    ( DatumHash
+    ( BabbageEra
+    , ConwayEra
+    , DatumHash
+    , IsRecentEra (..)
+    , RecentEra (..)
     , datumHashFromBytes
     )
 import Test.QuickCheck
     ( Gen
     , arbitrary
     , vectorOf
     )
+import Unsafe.Coerce
+    ( unsafeCoerce
+    )
+
+import Cardano.Ledger.Alonzo.PParams
+    ( OrdExUnits (..)
+    )
+import Cardano.Ledger.Babbage.PParams
+    ( BabbagePParams (..)
+    , CoinPerByte (..)
+    )
+import Cardano.Ledger.BaseTypes
+    ( BoundedRational (..)
+    , EpochInterval (..)
+    , ProtVer (..)
+    , natVersion
+    )
+import Cardano.Ledger.Conway.PParams
+    ( ConwayPParams
+    , DRepVotingThresholds (..)
+    , PoolVotingThresholds (..)
+    , upgradeConwayPParams
+    )
+import Cardano.Ledger.Plutus
+    ( CostModel
+    , ExUnits (..)
+    , Language (..)
+    , Prices (..)
+    , mkCostModel
+    , mkCostModels
+    )
+import Data.Functor.Identity
+    ( Identity
+    )
+import Data.Ratio
+    ( (%)
+    )
 
 import qualified Data.ByteString as BS
+import qualified Data.Map as Map
 
 genDatumHash :: Gen DatumHash
 genDatumHash =
     fromMaybe (error "genDatumHash should always generate valid hashes")
     . datumHashFromBytes
     . BS.pack <$> vectorOf 32 arbitrary
+
+--------------------------------------------------------------------------------
+-- PParams
+--------------------------------------------------------------------------------
+
+-- | We try to use similar parameters to mainnet where it matters (in particular
+-- fees, execution unit prices, and the cost model.)
+mockPParams
+    :: forall era . IsRecentEra era => PParams era
+mockPParams = case recentEra @era of
+    RecentEraBabbage -> unsafeWrap babbagePParams
+    RecentEraConway -> unsafeWrap conwayPParams
+  where
+    -- The constructor of the 'Cardano.Ledger.Core.PParams.PParams' newtype is
+    -- not exported.
+    unsafeWrap :: PParamsHKD Identity era -> PParams era
+    unsafeWrap = unsafeCoerce
```

Comment:
I prefered having the compiler enforce that we don't forget to set a field, unlike using `emptyPParams` with lens setters. Could also have used `aeson-qq`, but constructing the Haskell type gives faster feedback if something is wrong and may be more helpful for writing a generator.